ui: add a sessions page

jordanlewis opened the pull request:
This PR adds two new pages:

1. Active Sessions, a table of all active sessions in the database.
2. Session Details, a page with information about a single active
   session.

In addition, active sessions and queries are now terminatable from the
Active Sessions page.

Release note (admin ui change): add the session list and session detail
pages, and permit session and query termination from the UI.

Release justification: low risk, high benefit change to admin ui

Closes #52193.

Screenshots:

![image](https://user-images.githubusercontent.com/43821/88502972-35a4da00-cf9e-11ea-953d-b3ffff7b272e.png)

![image](https://user-images.githubusercontent.com/43821/88502982-3d647e80-cf9e-11ea-8801-d41461a635eb.png)

![image](https://user-images.githubusercontent.com/43821/88502986-42293280-cf9e-11ea-91b7-2435c129f158.png)

jordanlewis committed:
ui: add active sessions and details pages

This commit adds two new pages:

1. Active Sessions, a table of all active sessions in the database.
2. Session Details, a page with information about a single active
   session.

In addition, active sessions and queries are now terminatable from the
Active Sessions page.

Release note (admin ui change): add the session list and session detail
pages, and permit session and query termination from the UI.

jordanlewis committed:
sql: add more information to active sessions protos

Add some new information to the transaction, session and query
protobufs, and to the transactions internal tables.

Release note (sql change): add 3 new columns to the
crdb_internal.node_transactions and cluster_transactions tables:
num_stmts, num_retries, and num_auto_retries. These keep track of the
total number of statements executed on the transaction, the number of
retries encountered, and the number of automatic retries encountered.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51903)
<!-- Reviewable:end -->

awoods187 commented:
This is amazing! I'm really excited we are tracking this information in the WebUI now.

jordanlewis commented:
Here's a short video of the page in action: https://youtu.be/xPAtTGchPfo

jordanlewis commented:
Actions menu:

![image](https://user-images.githubusercontent.com/43821/89428117-0adb2400-d70a-11ea-93a8-522298c37523.png)

Details page (the button location and colors need help, will need the CSS 🚁 later :)

![image](https://user-images.githubusercontent.com/43821/89428187-1cbcc700-d70a-11ea-948a-e32130f6670a.png)

![image](https://user-images.githubusercontent.com/43821/89428205-21817b00-d70a-11ea-865c-15dfebb18e21.png)

jordanlewis committed:
ui: tolerate empty new execution stats

Release note: None

jordanlewis commented:
Okay, this is ready for a review!

dhartunian reviewed (commented):
Thanks for putting this up @jordanlewis. My comments are mostly small nits. Nothing blocking except maybe the question re: pagination.

> Reviewed 4 of 4 files at r1, 11 of 11 files at r2, 25 of 27 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-:-MEQ4O_E3FTuQbHartTn:b-db07ct)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jordanlewis, @nathanstilwell, and @solongordon)

___
*[pkg/ui/src/redux/sessions/sessionsSagas.tsx, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ6upn80_gZawDAXqU:-MEQ6upn80_gZawDAXqV:b-nspy0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/redux/sessions/sessionsSagas.tsx#L76)):*
> ```XML
> }
> 
> export function* terminateQuerySaga(action: PayloadAction<ICancelQueryRequest>) {
> ```

I still don't fully "get" how redux-saga works but I do like how this code reads! Really clear what's intended.
___
*[pkg/ui/src/views/sessions/sessionsPage.tsx, line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ8BuB58xALyZFbWjh:-MEQ8BuB58xALyZFbWji:bwr1er1) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/sessionsPage.tsx#L62)):*
> ```XML
>         ascending: false,
>       },
>       pagination: {
> ```

I'm unable to glean from the code where the pagination is being applied to the data in the table...am I missing something?
___
*[pkg/ui/src/views/sessions/sessionsPage.tsx, line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ7fQc3tXQuQInceS-:-MEQ7fQc3tXQuQInceS0:b-xb23wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/sessionsPage.tsx#L160)):*
> ```XML
>               label: "Learn more",
>               // TODO(jordan): point this to a more appropriate page.
>               buttonHref: "https://www.cockroachlabs.com/docs/stable/show-sessions.html",
> ```

See `pkg/ui/src/util/docs.ts` for docs link examples. We can generate the base of the docs URL there.
___
*[pkg/ui/src/views/sessions/terminateSessionModal.tsx, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQDcU-1E3ELfKunu9F:-MEQDcU06saHgJUgx_z4:bmwvbde) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/terminateSessionModal.tsx#L26)):*
> ```XML
> 
> // tslint:disable-next-line:variable-name
> const TerminateSessionModal = (props: TerminateSessionModalProps, ref: React.RefObject<TerminateSessionModalRef>) => {
> ```

This whole pattern is a bit rough to read second time around. Thanks for lifting it from the diagnostics modal for now. I can't help but feel like there's a simplification that's possible here

cc @nathanstilwell, the diagnostics modal stuff was built quite quickly so I'm sure the modal pattern it implements could use a second look. Feels weird to be using these "refs" but maybe it's a valid pattern. Not sure how CC does it.
___
*[pkg/ui/src/views/statements/statementsTableContent.tsx, line 228 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQCcRmBL8C3cYhxTjl:-MEQCcRn2Jp7UeJROrZp:b-dfswfm) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/statements/statementsTableContent.tsx#L228)):*
> ```XML
> // StatementLinkTarget returns the link to the relevant statement page, given
> // the input statement details.
> export const StatementLinkTarget = (props: StatementLinkProps) => {
> ```

Thanks for improving this component! 


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51903)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @nathanstilwell, and @solongordon)

___
*[pkg/ui/src/views/sessions/sessionDetails.tsx, line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQI10r2rQoFeKvHzPB:-MEQI10sE_Z5syn_zvHo:bk8shm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd7f8f98510f8ac639709b894db3a4ef5c5a255/pkg/ui/src/views/sessions/sessionDetails.tsx#L133)):*
> ```XML
>           <div>
>             <Link className={cx("back-link")} to={ "/sessions" }>
>               Back to Statements
> ```

should say Sessions here?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-:-MESsWWp4QtTnU0R6qQl:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jordanlewis, @nathanstilwell, and @solongordon)

___
*[pkg/ui/src/redux/sessions/sessionsSagas.tsx, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ6upn80_gZawDAXqU:-MEQp4hE4BzyrIM_NDCn:b8k79wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/redux/sessions/sessionsSagas.tsx#L76)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I still don't fully "get" how redux-saga works but I do like how this code reads! Really clear what's intended.
</blockquote></details>

You and me both!
___
*[pkg/ui/src/views/sessions/sessionDetails.tsx, line 133 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQI10r2rQoFeKvHzPB:-MESsUed3Uw8H4pccUhA:br21pzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd7f8f98510f8ac639709b894db3a4ef5c5a255/pkg/ui/src/views/sessions/sessionDetails.tsx#L133)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should say Sessions here?
</blockquote></details>

Nice catch, fixed.
___
*[pkg/ui/src/views/sessions/sessionsPage.tsx, line 62 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ8BuB58xALyZFbWjh:-MEQv3Gn3yDb0aXIIIWF:bvib957) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/sessionsPage.tsx#L62)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I'm unable to glean from the code where the pagination is being applied to the data in the table...am I missing something?
</blockquote></details>

No, I just failed at cargo culting :) Thank you for noticing, I added the remaining plumbing.
___
*[pkg/ui/src/views/sessions/sessionsPage.tsx, line 160 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQ7fQc3tXQuQInceS-:-MEQvORL0bPLlWnOd3RD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/sessionsPage.tsx#L160)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

See `pkg/ui/src/util/docs.ts` for docs link examples. We can generate the base of the docs URL there.
</blockquote></details>

Done.
___
*[pkg/ui/src/views/sessions/terminateSessionModal.tsx, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-MEQDcU-1E3ELfKunu9F:-MEQvWgf5CO0RcNktuVy:b7x5irh) ([raw file](https://github.com/cockroachdb/cockroach/blob/727534d5d92521d44b7dbe22f2de87615b17b027/pkg/ui/src/views/sessions/terminateSessionModal.tsx#L26)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This whole pattern is a bit rough to read second time around. Thanks for lifting it from the diagnostics modal for now. I can't help but feel like there's a simplification that's possible here

cc @nathanstilwell, the diagnostics modal stuff was built quite quickly so I'm sure the modal pattern it implements could use a second look. Feels weird to be using these "refs" but maybe it's a valid pattern. Not sure how CC does it.
</blockquote></details>

Yes, these refs are definitely interesting. I thought this is how one manages state. See the way I implemented the per-row actions popovers for a more interesting example - it definitely feels like a pattern, but an odd one.


<!-- Sent from Reviewable.io -->

jordanlewis committed:
server: list sessions doesn't need remote debug on

Previously, running the list sessions or list local sessions APIs would
require the remote debugging setting to be turned on. Now that we have
role-based access controls for viewing session activity, this is no
longer necessary.

Release note: None
Release justification: low-risk update to new functionality

jordanlewis commented:
Latest revision uses UUID format for the links, includes session ID in the session page, freezes the session page from updates, and fixes the remote debugging issues.

yuzefovich commented:
Just a drive-by comment (based on the dilemma Jordan posed in his stream): re: should the webpage of a session reload itself when that session exits or should the webpage be frozen until the user initiates the reload? Not sure if it was considered, but maybe a good alternative would be to add a pop-up warning that says something like "the session has finished, click `reload`" to reload the page and have a button or something. (Sorry If the sessions page already does something like this.)

solongordon commented on pkg/ui/src/views/sessions/sessionsTable.tsx:
```diff
@@ -0,0 +1,218 @@
+                    terminateQueryRef?.current?.showModalFor({
+                      query_id: session.active_queries[0].id,
+                      node_id: session.node_id.toString(),
+                      username: session.username,
```
This is confusing but the username here should actually be the user who is canceling the session, not the user whose session it is. (I believe this field exists because when this gets run as a SQL command, the request is always made as root.) The way you have it right now, a user with CANCELQUERY privileges can't cancel other users' sessions.

solongordon commented on pkg/ui/src/views/sessions/sessionsTable.tsx:
```diff
@@ -0,0 +1,218 @@
+                onClick={() => terminateSessionRef?.current?.showModalFor({
+                  session_id: session.id,
+                  node_id: session.node_id.toString(),
+                  username: session.username,
```
Ditto here.

solongordon commented on pkg/ui/src/views/sessions/sessionDetails.tsx:
```diff
@@ -0,0 +1,346 @@
+                onClick={() => this.terminateSessionRef?.current?.showModalFor({
+                  session_id: session.id,
+                  node_id: session.node_id.toString(),
+                  username: session.username,
```
Ditto here.

solongordon commented on pkg/ui/src/views/sessions/sessionDetails.tsx:
```diff
@@ -0,0 +1,346 @@
+                    this.terminateQueryRef?.current?.showModalFor({
+                      query_id: session.active_queries[0].id,
+                      node_id: session.node_id.toString(),
+                      username: session.username,
```
Ditto here.

jordanlewis commented:
Done.

solongordon reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51903)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jordanlewis, and @nathanstilwell)



<!-- Sent from Reviewable.io -->

solongordon reviewed (commented):
By the way, I don't want this to block merging but it seems unideal that if you don't have CANCELQUERY permissions, trying to cancel a query or a session gives you a generic error when you click the link. Would be nicer to either not show the link or explain that you don't have the right privileges.

Same goes for if a non-admin tries to cancel an admin query/session.

I'd say let's just open an issue to track it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51903#-:-MGKkWjOB5cGZYl_utoX:b5kneqv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jordanlewis, and @nathanstilwell)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
Thanks! That piece of feedback is already in the acceptance testing doc as it turns out: https://docs.google.com/document/d/1obahPF5wH_GbScPFtA3VUz-UjgqR7IRMCKRuT0JFen8/edit# I'll merge this now and make issues next.

bors r+

jordanlewis commented:
bors r+

jordanlewis commented:
bors r-

jordanlewis commented:
bors r+

jordanlewis commented:
bors r+

🤞
